prow.sh

```diff
@@ -0,0 +1,902 @@
+            kubectl logs "$@" "$pod" "$container" >>"${ARTIFACTS}/$pod/$container.log" &
+            echo "$!"
+        done
+    done
+}
+
+# Makes the E2E test suite binary available as "${CSI_PROW_WORK}/e2e.test".
+install_e2e () {
+    if [ -e "${CSI_PROW_WORK}/e2e.test" ]; then
+        return
+    fi
+
+    git_checkout "${CSI_PROW_E2E_REPO}" "${GOPATH}/src/${CSI_PROW_E2E_IMPORT_PATH}" "${CSI_PROW_E2E_VERSION}" --depth=1 &&
+    if [ "${CSI_PROW_E2E_IMPORT_PATH}" = "k8s.io/kubernetes" ]; then
+        run_with_go "${CSI_PROW_GO_VERSION_E2E}" make WHAT=test/e2e/e2e.test "-C${GOPATH}/src/${CSI_PROW_E2E_IMPORT_PATH}" &&
+        ln -s "${GOPATH}/src/${CSI_PROW_E2E_IMPORT_PATH}/_output/bin/e2e.test" "${CSI_PROW_WORK}"
+    else
+        run_with_go "${CSI_PROW_GO_VERSION_E2E}" go test -c -o "${CSI_PROW_WORK}/e2e.test" "${CSI_PROW_E2E_IMPORT_PATH}/test/e2e"
+    fi
+}
+
+# Makes the csi-sanity test suite binary available as
+# "${CSI_PROW_WORK}/csi-sanity".
+install_sanity () (
+    if [ -e "${CSI_PROW_WORK}/csi-sanity" ]; then
+        return
+    fi
+
+    git_checkout "${CSI_PROW_SANITY_REPO}" "${GOPATH}/src/${CSI_PROW_SANITY_IMPORT_PATH}" "${CSI_PROW_SANITY_VERSION}" --depth=1 || die "checking out csi-sanity failed"
+    run_with_go "${CSI_PROW_GO_VERSION_SANITY}" go test -c -o "${CSI_PROW_WORK}/csi-sanity" "${CSI_PROW_SANITY_IMPORT_PATH}/cmd/csi-sanity" || die "building csi-sanity failed"
+)
+
+# Whether the hostpath driver supports raw block devices depends on which version
+# we are testing. It would be much nicer if we could determine that by querying the
+# installed driver.
+hostpath_supports_block () {
+    if [ -e "cmd/hostpathplugin" ] && ${CSI_PROW_BUILD_JOB}; then
+        # The assumption is that if we build the hostpath driver, then it is
+        # a current version with support.
+        echo true
+        return
+    fi
+
+    case "${CSI_PROW_DEPLOYMENT}" in kubernetes-1.13) echo false;; # wasn't supported and probably won't be backported
+                                      *) echo true;; # probably all other deployments have a recent driver
+    esac
+}
+
+# Captures pod output while running some other command.
+run_with_loggers () (
+    loggers=$(start_loggers -f)
+    trap 'kill $loggers' EXIT
+
+    run "$@"
+)
+
+# Invokes the filter-junit.go tool.
+run_filter_junit () {
+    run_with_go "${CSI_PROW_GO_VERSION_BUILD}" go run "${RELEASE_TOOLS_ROOT}/filter-junit.go" "$@"
+}
+
+# Runs the E2E test suite in a sub-shell.
+run_e2e () (
+    name="$1"
+    shift
+
+    install_e2e || die "building e2e.test failed"
+    install_ginkgo || die "installing ginkgo failed"
+
+    # TODO (?): multi-node cluster (depends on https://github.com/kubernetes-csi/csi-driver-host-path/pull/14).
+    # When running on a multi-node cluster, we need to figure out where the
+    # hostpath driver was deployed and set ClientNodeName accordingly.
+
+    # The content of this file depends on both what the E2E suite expects and
```

Comment:
Can this go in a subdirectory with the per version hostpath deployment files?

Reply:
Not really, because it's content also depends on the E2E suite, and the example deployment files know nothing about that.

Reply:
What if the hostpath driver keeps multiple versions of its test config depending on the e2e test version?  I want to move as much driver-specific, or job-specific configuration out of this script.

Reply:
The E2E test suite could be a certain revision instead of a tagged version, so the mechanism for picking a certain file from the csi-driver-host-path repo or a script inside that repo which outputs the content could become quite complicated.

Keeping it here has the advantage that the test driver config can be updated in lockstep with updating the `CSI_PROW_E2E` variables.

I would prefer to defer this "move configuration out of this script".